x-pack/plugins/reporting/server/lib/task.ts

```diff
@@ -0,0 +1,288 @@
+    }
+
+    return docOutput;
+  }
+
+  private async _performJob(job: TaskParams<unknown>) {
+    if (!this.taskExecutors) {
+      throw new Error(`Task run function factories have not been called yet!`);
+    }
+
+    const { report, payload } = job;
+    logRunner(this.logger, 'info', `Starting job ${report.jobtype} job ${report.id}`);
+
+    // get the run_task function
+    const runner = this.taskExecutors.get(report.jobtype);
+    if (!runner) {
+      throw new Error(`No defined task runner function for ${report.jobtype}!`);
+    }
+
+    // return the run output
+    const cancellationToken = new CancellationToken(); // FIXME: allow to be cancellable
+    return await runner(report.id, payload, cancellationToken);
+  }
+
+  private async _completeJob(job: TaskParams<unknown>, output: TaskRunResult) {
+    const { report: taskJson } = job;
+    logRunner(this.logger, 'info', `Saving ${taskJson.jobtype} job ${taskJson.id}`);
+    const completedTime = moment().toISOString();
+    const docOutput = this._formatOutput(output);
+
+    const store = this.getStore();
+    const doc = {
+      completed_at: completedTime,
+      output: docOutput,
+    };
+    const report = await store.findReport(taskJson);
+    await store.setReportCompleted(report, doc);
+  }
+
+  public register(taskManager: TaskManagerSetupContract) {
+    taskManager.registerTaskDefinitions({
+      [REPORTING_QUEUE_JOB_TYPE]: {
+        type: REPORTING_QUEUE_JOB_TYPE,
+        title: 'Reporting job queue',
+        createTaskRunner: this.createTaskRunner,
+      },
+    });
+  }
+
+  /*
+   * Reporting job scheduler
+   */
+  public scheduleTaskFactory(taskManager: TaskManagerStartContract): SchedulingFn {
+    const scheduleTask = async <TaskParamsType>({
+      report,
+      payload,
+      user,
+      headers,
+    }: TaskParams<TaskParamsType>) => {
+      const taskInstance: ReportingTaskInstanceParams<TaskParamsType> = {
+        taskType: REPORTING_QUEUE_JOB_TYPE,
+        state: {},
+        params: { report, payload, user, headers },
+      };
+
+      return await taskManager.schedule(taskInstance);
+    };
+
+    return scheduleTask;
+  }
+
+  /*
+   * Reporting job worker
+   */
+  private createTaskRunner = (context: RunContext) => ({
+    run: async () => {
+      // find the job in the store and set status to processing
+      const job = context.taskInstance.params as TaskParams<unknown>;
+
+      try {
+        await this._claimJob(job);
+      } catch (err) {
+        this.logger.error(`error in claimJob! ${job.report.id}`);
+        this.logger.error(err);
+        throw err;
+      }
+
+      let output: TaskRunResult;
+      try {
+        // perform job
+        output = await this._performJob(job);
+        // update the report in the store
+        await this._completeJob(job, output);
+      } catch (error) {
+        await this._failJob(job, error);
+      }
+
+      // return back to Task Manager
+      // TODO return errors
```

Comment:
This TODO has to be worked on. Without errors returned back to task manager, there will never be retries.